Monix streaming with NDBC

tschuchortdev opened the pull request:
I'm attempting to write a Monix streaming context using NDBC based on the existing code for JDBC. This is an early WIP PR for visibility. So far I have extracted logic from the `NdbcContext` to a `NdbcContextBase` class similar to how it was done in the JDBC modules but replaced the `ContextEffect` class with the appropriate Cats typeclass because I find it much more readable and we get lawful instances for many effect types (like Monix `Task`) for free. 
If you have any hints or suggestions I would be very thankful since I'm not that well versed in databases.

@getquill/maintainers

tschuchortdev commented:
I removed the cats typeclasses again because I read in another issue that some of Quill's users don't allow cats and it turned out that we can't write lawful instances for `Future` anyway. I also added a whole bunch of  `TODO` comments with questions and remarks. If anyone could take a look at them it would be much appreciated.

tschuchortdev commented:
Tests for NDBC Monix stuff are now working, although NDBC apparently doesn't support transactions with streaming, so I had to ignore those tests. For some reason JDBC tests are failing with connection issues even though I barely changed anything. Also there is a nasty issue with macros when the type of a `Context` member variable is narrowed in a deep hierarchy of mixin interfaces (see `PeopleMonixSpec` and `PeopleSpec` in `quill-jdbc-monix` module). For some reason this doesn't work with the NDBC contexts. I worked around it by pasting the code from `PeopleMonixSpec` into `PeopleMonixNdbcSpec` and inherting from `PeopleSpec` directly.

tschuchortdev commented on quill-core/src/main/scala/io/getquill/context/ContextEffect.scala:
```diff
@@ -16,11 +16,11 @@ trait ContextEffect[F[_]] {
    * Map a parameter of the effect. This is really just a functor.
    */
-  def push[A, B](result: F[A])(f: A => B): F[B]
+  def fmap[A, B](result: F[A])(f: A => B): F[B]
```
I mostly changed it to make understanding for myself easier, but I can change it back if you want.

tschuchortdev commented:
If I can find the time I will try to have a look this weekend and update the PR to make it more manageable. Some tests are still failing for no apparent reason, do you have any idea what the issue might be? 
> For some reason JDBC tests are failing with connection issues even though I barely changed anything

juliano commented:
> If I can find the time I will try to have a look this weekend and update the PR to make it more manageable. Some tests are still failing for no apparent reason, do you have any idea what the issue might be?

Can you please rebase it, so I can have a look?

tschuchortdev commented:
@juliano I have rebased it and cleaned up a bit, but SBT is refusing to run the tests because of an "unused import" in quill-core that is not unused at all.

juliano commented:
Which ones are failing? Besides, you've probably seen the [big change related to Quats](https://github.com/getquill/quill/pull/1911), so another rebase will be necessary. And if you can remove the import conflicts, this one will probably be ready :)

tschuchortdev commented on quill-cassandra-monix/src/main/scala/io/getquill/CassandraMonixContext.scala:
```diff
@@ -22,9 +22,6 @@ class CassandraMonixContext[N <: NamingStrategy](
   with MonixContext[CqlIdiom, N] {
 
-  // not using this here
-  override val effect = Runner.default
```
Why? It's no longer required to be overwritten after moving `Runner` to `MonixJdbcContext` (which was required to be overwritten from `JdbcContextBase` but `CassandraMonixContext` is not a JDBC context obviously).

tschuchortdev commented:
Should be working now. Locally I had 4 tests fail, but I think that's just some peculiarities of my machine:

```
[error] (quill-async-postgres / Test / testQuick) sbt.TestsFailedException: Tests unsuccessful
[error] (quill-jasync-postgres / Test / testQuick) sbt.TestsFailedException: Tests unsuccessful
[error] (quill-coreJS / Test / fastOptJS) There were linking errors
[error] (quill-sqlJS / Test / fastOptJS) There were linking errors
```

The top 2 have to do with date handling:
```
[info] - Java8 times *** FAILED ***
[info]   Vector(2020-07-26T19:11:36.020) was not equal to List(2020-07-26T19:11:36.020795) (ArrayAsyncEncodingSpec.scala:50)
```
This is on MacOS with OpenJDK 13.0.1. Perhaps that's worth investigating?

tschuchortdev commented on quill-monix/src/main/scala/io/getquill/context/monix/MonixContext.scala:
```diff
@@ -16,6 +16,4 @@ trait MonixContext[Idiom <: io.getquill.idiom.Idiom, Naming <: NamingStrategy] e
   def executeQuery[T](sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): Task[List[T]]
   def executeQuerySingle[T](sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): Task[T]
-
-  protected val effect: Runner
```
@juliano Regarding the `CassandraMonixContext`: Perhaps removing this was not ok? Then the `CassandraMonixContext` would also need to implement `effect` like it did before. But both were unused or I wouldn't have removed them.

juliano replied:
After reviewing again I don't think it's a problem 👍

juliano commented on quill-ndbc/src/test/scala/io/getquill/context/ndbc/NdbcContextEffectSpec.scala:
```diff
@@ -1,31 +0,0 @@
-import io.getquill.Spec
-import io.trane.future.scala.{ Await, Future }
-
-class NdbcContextEffectSpec extends Spec {
```
Why removing this test?

tschuchortdev replied:
Good question... I think I removed it because I also removed the NdbcContextEffect, so there should be a test for the replacement, but maybe I forgot to replace the test. I will check later.

juliano commented on quill-ndbc/src/main/scala/io/getquill/context/ndbc/NdbcContext.scala:
```diff
@@ -37,107 +22,72 @@ abstract class NdbcContext[I <: SqlIdiom, N <: NamingStrategy, P <: PreparedStat
-      prepare(createPreparedStatement(statement))._1.reverse.map(prepareParam)
+      /* TODO: I'm not sure if it's a good idea to use wrap here. The whole thing with `wrap` and
+          `withDataSource` is a mess and IMO should be cleaned up to be simpler and more compositional
+       */
```
Would you mind changing this comment? It sounds harsh to call the work that's been done so far a mess.

tschuchortdev replied:
I'm going to remove all the TODO comments before merging. To be clear this wasn't meant in a bad way. I'm sure there's a reason why it was done this way but I felt I needed to point out that it feels unintuitive as a reader without the background knowledge.

juliano commented:
@tschuchortdev I've added a couple of comments, other than that seems good to go. Thanks for your work here!

tschuchortdev commented on quill-ndbc/src/main/scala/io/getquill/context/ndbc/NdbcContextBase.scala:
```diff
@@ -0,0 +1,142 @@
+      }
+    )(_.flatten)
+
+  // TODO: Should this be blocking? Previously it was just a Future wrapped in a Try, which makes no sense
+  def probe(sql: String): Try[_] =
+    Try(runBlocking(withDataSourceFromFuture(_.query(sql).toScala), Duration.Inf))
```
^

tschuchortdev commented on quill-ndbc/src/main/scala/io/getquill/context/ndbc/NdbcContextBase.scala:
```diff
@@ -0,0 +1,142 @@
+    else
+      acc.reverse
+
+  def transaction[T](f: => Result[T]): Result[T] = withDataSource { ds =>
+    /* TODO: I'm assuming that we don't need to turn autocommit off/on for streaming because I can't
+        find any way to do so with the NDBC DataSource and it seems to handle streaming on its own */
+
+    implicit def javaSupplier[S](s: => S): Supplier[S] = new Supplier[S] {
+      override def get = s
+    }
+
+    val javaFuturePool = FuturePool.apply(Executors.newCachedThreadPool())
+
+    resultEffect.fromDeferredFuture(implicit scheduler =>
+      javaFuturePool.isolate(
+        ds.transactional {
+          resultEffect.toFuture(f, scheduler).toJava
+        }
+      ))
+  }
```
^

tschuchortdev commented on quill-ndbc-monix/src/main/scala/io/getquill/context/monix/MonixNdbcContext.scala:
```diff
@@ -0,0 +1,149 @@
+  protected def withDataSourceObservable[T](f: DataSource[P, R] => Observable[T]): Observable[T] =
+    schedule(f(dataSource))
+
+  // TODO: What about fetchSize? Not really applicable here
+  def streamQuery[T](fetchSize: Option[Index], sql: String, prepare: Prepare = identityPrepare, extractor: Extractor[T] = identityExtractor): Observable[T] =
+    Observable
+      .eval {
+        // TODO: Do we need to set ResultSet.TYPE_FORWARD_ONLY, ResultSet.CONCUR_READ_ONLY?
+        val stmt = createPreparedStatement(sql)
+        val (params, ps) = prepare(stmt)
+        logger.logQuery(sql, params)
+        ps
+      }
+      .flatMap(ps => withDataSourceObservable { ds =>
+        Observable.fromReactivePublisher(ds.stream(ps))
+      })
+      .map(extractor)
```
^

tschuchortdev commented on quill-ndbc/src/main/scala/io/getquill/context/ndbc/NdbcContext.scala:
```diff
@@ -37,107 +22,72 @@ abstract class NdbcContext[I <: SqlIdiom, N <: NamingStrategy, P <: PreparedStat
-    ).toScala
+  override def transaction[T](f: => Future[T]): Future[T] = super.transaction(f)
+
+  /* TODO: I'm assuming that we don't need to bracket and close the dataSource like with JDBC
+      because previously it wasn't done here either */
+  override def withDataSource[T](f: DataSource[P, R] => Future[T]): Future[T] = f(dataSource)
```
^

tschuchortdev commented:
Alright, I've added tests for the `ContextEffect` types that were missing and deleted unnecessary TODO comments, but I left some TODO comments in that I felt would be helpful in the future/are unanswered and marked them here in the code review. From my end this is ready to merge unless you have any further requests for improvements.

juliano commented:
Seems good to go! Can you please rebase it one last time?

tschuchortdev commented:
Done
